### DISCUSSION TITLE
Not parser for keywords and a deprecation question

### DISCUSSION THREAD

**Discussion by Isaac-Leonard** [zesterer/chumsky#105] [Ideas]
Would it be possible to add a not parser so we could do something like:
```
let assign = ident()
   .not(seq("var".chars())
    .then(just('=').padded())
    .then_ignore(exp);
```

So we could implement key words for our languages?.
I quickly looked at the filter parser but couldn't see how to make it really work for this.

Also as an aside I've been getting deprecation warnings about 'seq()', but couldn't replace it with 'just()' for strings, how do I do this properly?
Best regards Isaac

**Comment by zesterer** [comment]
Hello,

> So we could implement key words for our languages?.

This sounds like you want to try parsing declarations before assignments rather than trying to specifically exclude declarations from the assignment parser. Something like this:

```rs
let decl = text::keyword("var")
    .then(text::ident())
    .then_ignore(just('=').padded());

let assign = ident()
    .then(just('=').padded())
    .then_ignore(exp);

decl.or(assign)
```

That aside, it seems like you're parsing at the same time as lexing. Although this is viable, it does mean that you're likely to end up with `.padded()` appearing in a lot of weird and inconvenient places. I would recommend separating lexing and parsing out into distinct passes, as the `nano_rust` example does.

> Also as an aside I've been getting deprecation warnings about 'seq()', but couldn't replace it with 'just()' for strings, how do I do this properly?

`just("my string")`
`just('c')`
`just(String::from("my string"))`
`just(['s', 't', 'r', 'i', 'n', 'g'])`
`just(vec!['s', 't', 'r', 'i', 'n', 'g'])`

Any of these should work with `just`.

Happy to clarify if you still have issues!

**Comment by Isaac-Leonard** [comment]
Oh wow, I've somehow managed to miss the keyword function this whole time lol, just also seems to work now though I'm not don't know what I'm doing different then before lol.
Thank you

**Comment by Isaac-Leonard** [comment]
I think I worked out what I was doing wrong with 'just()'
It seems to not like receiving a reference to owned strings and borrowing the result of a function call such as
```
just(&get_string())
```
gives a temporary value is dropped error, but removing the borrow makes it work fine.
